WIP: Glossary and guide to referencing

handrews opened the pull request:
This includes both a defitions page (which could be renamed to glossary if folks like that better) and the start of a detailed guide to referencing and related topics.

The referencing guide is more of a WIP than the glossary, and I'm happy to split the glossary into its own PR if folks like it enough to approve it.  I just wanted to show it in context as this is how I wrote it all.

I made the referencing information a section as it is too long and has too many sub-topics to be on a single page.  I tried to extract the most common topics into the FAQ so that the basics can be found quickly, which might lead some folks to read the more deep-dive pages.

Some pages stop rather abruptly as I was getting bogged down and figured it would be good to solicit feedback before writing too much more.  I'm happy to rework any/all of this to suit what typical users want to know.

duncanbeevers commented on referencing/faq.md:
```diff
@@ -0,0 +1,74 @@
+
+In general, no.  But in many common cases, yes, and tools exist that will attempt this.
+
+There are two types of reasons why de-referenceing might not be possible:
```
~de-referenceing~ de-referencing

duncanbeevers commented on referencing/faq.md:
```diff
@@ -0,0 +1,74 @@
+There are two types of reasons why de-referenceing might not be possible:
+
+* Technical reasons:
+   * Cyclic references are allowed in Schema Objects, and are used to describe recursive tree structures.  JSON Schema itself is a recursive tree structure, so the JSON Schema metaschemas cannot be fully de-referenced (see draft-07 or earlier for how this works with `"$ref"`; draft 2020-12 uses a different technique)
```
This point about recursive references is important, and implies we *can't* ship a completely normalized document (all references resolved and inlined), and that some level of at least intra-document indirection must be supported. (eg; tools must be *slightly* more complicated than simple JSON / YAML parsers)

notEthan reviewed (commented):
This is very good, quite informative, thank you for writing it. It clarified several things I didn't realize I had misconceptions about.

handrews commented on referencing/faq.md:
```diff
@@ -0,0 +1,74 @@
+
+## Do I need to understand dynamic references?
+
+This guide does **not** cover OAS 3.1 / JSON Schema draft 2020-12's dynamic references (`"$dynamicRef"` and `"$dynamicAnchor"`).  Dynamic references seve a different purpose than  discussed here.  The following JSON Schema Blog posts explain some uses of dynamic references and anchors:
```
Now I want to make up a meaning for "seve" where this sentence actually makes sense... 🙃

handrews commented on referencing/what.md:
```diff
@@ -0,0 +1,61 @@
+
+## Non-reference linkages
+
+The following connections between parts of the OAS require that some name or identifier has been seen.  Due to ambiguities in the specification (which the OAI will hopefully be able to clarify in future patch releases), it is not always clear how to handle identifiers that are present in a _document_ that has been referenced, but are not present in the _exact JSON object_ targeted by any single r ...
```
This is great feedback, thanks!  I will think on this.  Maybe I should just add an example here (there might be one in my [detailed processing model report](https://github.com/OAI/oascomply/blob/main/reports/processing-model-details.md) that's suitable).

handrews commented on referencing/what.md:
```diff
@@ -0,0 +1,61 @@
+| 3.x         | Components Object  | Security Scheme name | Security Requirement Object |
+| 3.x         | Tag Object         | `name`            | Operation Object |
+| 3.x         | Operation Object   | `operationId`     | Link Object |
+| 3.x         | Server Object      | n/a               | Paths Object |
```
Actually, in my notes Parameter Object and Server Variable Object are included as linkages.  I left them out because they felt different, mostly because they connect things that are close together.  But come to think of it, the Parameters Object can go in two places, and be `$ref`'d, so it's probably good to include.  And might as well include the Server Variable Object even if it is more localized.  There's also the `parameters` field in the Link Object which I'm not sure I noticed before now.  I'll add all of these.

Feel free to go hunting!  There might be more hiding somewhere, it's a big spec :-)

karenetheridge commented on referencing/faq.md:
```diff
@@ -0,0 +1,74 @@
+1. consists only of a single JSON or YAML document
+1. contains only references that are fragments (beginning with a `"#"`, e.g. `{"$ref": "#/components/responses/oops"}`)
+1. does not have any other keywords in the same object as any `"$ref"`
+1. (OAS 3.1 only) does *not* use `"$id"`if you are using OAS 3.1, you are  *not* using the `"$id"` keyword Schema Objects
```
```suggestion
1. (OAS 3.1 only) does *not* use `"$id"`if you are using OAS 3.1, you are  *not* using the `"$id"` keyword in Schema Objects
```

handrews commented:
@duncanbeevers @notEthan @karenetheridge @ndenny I think I have incorporated all feedback, please let me know if I missed something.

@notEthan instead of making bullet points, I rearranged and renamed the table columns- please let me know if that is more readable.  I'd like to keep all of those sorts of blocks of information in the same sort of table rather than switching formats if possible.

I also added an example of names that look usable but probably won't be in many tools because the names are not directly referenced.  It's such a weird thing that I'm not sure if that makes it less or more confusing, but it is at least concrete.

duncanbeevers commented on referencing/faq.md:
```diff
@@ -0,0 +1,74 @@
+   * Cyclic references can also occur when describing complex payloads such as `multipart/form-data` where media type descriptions can be nested
+   * References in Link Objects (`"operationRef"`) and Discriminator Objects (`"mapping"`) cannot be replaced with literal values, although there are alternatives to using references in these cases (`"operationId"` and mapping by schema component name, respectively, althouogh there can be challenges with arranging the documents properly for those techniques as well)
+   * Dynamic references in OAS 3.1 (which are not covered in detail by this guide) have runtime behavior which cannot always be predicted ahead of time
+* Social reasons:
```
I didn't totally understand the objections in today's discussion, but I felt perhaps there was a desire to acknowledge these concerns, while not putting them front-and-center.

Rather than a rich breakdown of each of these *valid* social concerns, an extremely terse caveat could cover the same terrain from a slightly-higher elevation.

```suggestion
* Social reasons:
  Complex OpenAPI Descriptions may have many stakeholders, and be deployed and consumed in many contexts. Authors of OpenAPI Descriptions may need to consider these contexts when de-referencing.
  * Different owners; parts of an API may be owned by different teams, and may be deployed on different schedules
  * Security Policies; parts of an API may not be accessible for de-referencing by all consumers
```

In the service of brevity, I sort of squashed-together the `Large-scale` and `Independent Deployment` bits of the original bullet points.

The `Standards Compliance` piece is the bullet I understand the least (in terms of how an OAD might need to be split-up), but if there's a way to describe it succinctly, or to roll it into one of the other bullets, that seems like a win.

Again, I'm just trying to respond to what felt like a desire to downplay the (over)-articulation of these caveats, and to highlight the *Happy Path* for casual readers of these documents.

The concerns seem valid, but those who are affected by those concerns will likely have their ears pricked by the presence of the keywords, and don't need their concerns articulated in this context.

earth2marsh replied:
I like these suggestions. 

My general point in these discussions has been to observe that in a compile-time contract context, it might be helpful to publish a version of the description that is optimized for the consumer's point of view. In this case, you might consider doing things like merging multiple documents into a single one, and/or you might exclude operations to which the consumer was not entitled. In this way, the points about different owners and different security polices would be opaque (dynamic, just-in-time contract probably wouldn't ever fit this simplest of contexts).

handrews commented:
Closing this for now in favor of #79 for the definitions/glossary and doing more work on the references parts before re-submitting.
